core/server/worker/src/main/java/alluxio/worker/grpc/AbstractReadHandler.java

```diff
@@ -126,11 +153,46 @@ public void onNext(alluxio.grpc.ReadRequest request) {
       mContext.setPosReceived(mContext.getRequest().getStart());
       mDataReaderExecutor.submit(createDataReader(mContext, mResponseObserver));
       mContext.setDataReaderActive(true);
+    } catch (RejectedExecutionException e) {
+      handleStreamEndingException(e, Status.RESOURCE_EXHAUSTED);
     } catch (Exception e) {
-      LogUtils.warnWithException(LOG, "Exception occurred while processing read request {}.",
-          request, e);
-      mSerializingExecutor.execute(() -> mResponseObserver
-          .onError(AlluxioStatusException.fromCheckedException(e).toGrpcStatusException()));
+      handleStreamEndingException(
+          AlluxioStatusException.fromThrowable(e).toGrpcStatusException());
+    }
+  }
+
+  private void handleStreamEndingException(StatusException e) {
+    handleStreamEndingException(e, e.getStatus());
+  }
+
+  /**
+   * Handle any exception which should abort the client's read request.
+   *
+   * @param exception the exception thrown
+   * @param status the type of {@link Status} exception which should be returned to the user
+   */
+  private void handleStreamEndingException(Exception exception, Status status) {
+    Long sessionId = mContext.getRequest() == null ? null : mContext.getRequest().getSessionId();
+    LogUtils.warnWithException(LOG, "fatal error occurred while handling read. sessionId: {}",
+        sessionId, exception);
+    AlluxioStatusException statusExc = new AlluxioStatusException(status);
+    try (LockResource lr = new LockResource(mLock)) {
+      if (mContext == null) {
```

Comment:
We can, but we should still create one, because in the case that we want to respond with an error, and there are packets in flight that could trigger the handler's `onNext`, then it might still try to run the data reader even though we've already sent the error response.

Reply:
I see. That will never happen in current logic unless the client is hacked. I am fine with being extra cautious though.